File: src/main/java/me/dreamhopping/pml/api/PufferfishAPI.java
package me.dreamhopping.pml.api;

import me.dreamhopping.pml.api.provider.PufferfishAPIProvider;
import me.dreamhopping.pml.mods.Mod;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class PufferfishAPI implements Mod {
    public static final Logger LOGGER = LogManager.getLogger("PufferfishAPI");

    public void initialize() {
        LOGGER.info("Waking up PufferfishAPIProvider");

        try {
            Class.forName("me.dreamhopping.pml.api.impl.PufferfishAPIProviderImpl", true, this.getClass().getClassLoader());
        } catch (ClassNotFoundException e) {
            System.err.println("PufferfishAPIProviderImpl doesn't want to wake up... They seemed pretty angry... This is what they said to me:");
            e.printStackTrace();

            System.exit(-1);
        }

        // Do some required things like transformers, class bindings, etc.
        PufferfishAPIProvider.getInstance().applyTransformers();
        PufferfishAPIProvider.getInstance().applyBindings();
    }
}

File: src/main/java/me/dreamhopping/pml/api/PufferfishAPI.kt
package me.dreamhopping.pml.api

import me.dreamhopping.pml.api.bindings.provider.PufferfishAPIProvider
import me.dreamhopping.pml.mods.Mod
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import kotlin.system.exitProcess

class PufferfishAPI : Mod {
    val logger: Logger = LogManager.getLogger("PufferfishAPI")

    override fun initialize() {
        logger.info("Waking up PufferfishAPIProvider")

        try {
            Class.forName("me.dreamhopping.pml.api.impl.PufferfishAPIProviderImpl", true, this.javaClass.classLoader)
        } catch (e: Exception) {
            logger.error(
                "PufferfishAPIProviderImpl doesn't want to wake up... They seemed pretty angry... This is what they said:",
                e
            )
            exitProcess(-1)
        }

        // Do some required things like transformers, class bindings, etc.
        PufferfishAPIProvider.applyTransformers()
        PufferfishAPIProvider.applyBindings()
    }

    override fun unload() {
        // TODO: Maybe do some things
    }
}